delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/file/Files.scala

```diff
@@ -0,0 +1,119 @@
+package ch.epfl.bluebrain.nexus.delta.plugins.file
+
+import cats.effect.Clock
+import ch.epfl.bluebrain.nexus.delta.plugins.file.model.Digest.NotComputedDigest
+import ch.epfl.bluebrain.nexus.delta.plugins.file.model.FileCommand._
+import ch.epfl.bluebrain.nexus.delta.plugins.file.model.FileEvent._
+import ch.epfl.bluebrain.nexus.delta.plugins.file.model.FileRejection._
+import ch.epfl.bluebrain.nexus.delta.plugins.file.model.FileState.{Current, Initial}
+import ch.epfl.bluebrain.nexus.delta.plugins.file.model.{FileCommand, FileEvent, FileRejection, FileState}
+import ch.epfl.bluebrain.nexus.delta.sdk.utils.IOUtils
+import monix.bio.{IO, UIO}
+
+class Files {}
+
+object Files {
+
+  /**
+    * The files module type.
+    */
+  final val moduleType: String = "file"
+
+  private[file] def next(
+      state: FileState,
+      event: FileEvent
+  ): FileState = {
+    // format: off
+    def created(e: FileCreated): FileState = state match {
+      case Initial     => Current(e.id, e.project, e.storage, e.attributes, Map.empty, e.rev, deprecated = false,  e.instant, e.subject, e.instant, e.subject)
+      case s: Current  => s
+    }
+
+    def updated(e: FileUpdated): FileState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, storage = e.storage, attributes = e.attributes, updatedAt = e.instant, updatedBy = e.subject)
+    }
+
+    def updatedAttributes(e: FileComputedAttributesUpdated): FileState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, storage = e.storage, attributes = s.attributes.copy( mediaType = e.mediaType, bytes = e.bytes, digest = e.digest), updatedAt = e.instant, updatedBy = e.subject)
+    }
+
+    def tagAdded(e: FileTagAdded): FileState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, tags = s.tags + (e.tag -> e.targetRev), updatedAt = e.instant, updatedBy = e.subject)
+    }
+    // format: on
+
+    def deprecated(e: FileDeprecated): FileState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, deprecated = true, updatedAt = e.instant, updatedBy = e.subject)
+    }
+
+    event match {
+      case e: FileCreated                   => created(e)
+      case e: FileUpdated                   => updated(e)
+      case e: FileComputedAttributesUpdated => updatedAttributes(e)
+      case e: FileTagAdded                  => tagAdded(e)
+      case e: FileDeprecated                => deprecated(e)
+    }
+  }
+
+  private[file] def evaluate(
```

Comment:
Where do we check the right to write on the storage / the size of the file ?
It should be here no ?

Reply:
For files I'm afraid the routes are going to have this logic. Most likely in the routes we need to fetch the storage because we need it for the directive `withSizeLimit(storage.maxFileSize)`. And since we have the storage, we also check the permissions in the routes:
```scala
withSizeLimit(storage.maxFileSize)  & authorizeFor(projectPath, storage.writePermission) {
   fileUpload("file") { case (metadata, byteSource) =>
      // call the files bundle here
   }
}
```

If you guys have other ideas about how to do this in the `Files` stm without a big port of akka http functionality from those directives, I'm willing to do it inside `Files` stm

Reply:
If it is a stream, you might not always know the size of the file having the `HttpEntity`, since it can be chuncked. The size in that case is computed. So yeah I don't see an easy and clean way to do it inside `Files`. it seems much cleaner to do it in the routes in this case, as shown in the example.

Reply:
It seems to be done automatically by throwing an exception when you read the stream:
https://github.com/akka/akka-http/blob/master/akka-http-core/src/main/java/akka/http/javadsl/model/HttpEntity.java#L102

Reply:
Yes, but while [consuming the stream](https://github.com/akka/akka-http/blob/04ccca8d5e622f13f1711fc3f08a3581a6073293/akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala#L631), which is when calling `fileUpload`